lucet-runtime/lucet-runtime-internals/src/module/dl.rs

```diff
@@ -117,13 +120,113 @@ impl DlModule {
             }
         };
 
+        if module_data.import_functions().len() > 0 {
+            DlModule::resolve_imports(&module_data, function_manifest)?;
+        }
+
         Ok(Arc::new(DlModule {
             lib,
             fbase,
             module_data,
             function_manifest,
         }))
     }
+
+    /// `resolve_imports` fixes up entries in function manifest for imported functions,
+    /// which initialize to 0, with the correct address of the referenced function
```

Comment:
I don't understand why we couldn't do the same thing by using a relocation and having the loader take care of this. In `lucetc::function_manifest`, Can we get rid of the `len > 0` comparison in `write_relocated_slice` so that we emit a relocation to the import function symbol? The import function has relocations that work at call sites, so I think it should be possible for it to work here as well.

Maybe there's some reason for this that I haven't thought of. Otherwise this looks like good code to me.

Reply:
just realized that import funcs wont have a trap table, so we'd still need to handle that case properly. we could provide the `Linkage` of a func in the `function_manifest` vec to make it more clear what is going on there.

Reply:
Using a relocation for these is a good idea. The `len > 0` check was to avoid writing a relocation to a blob of data we wouldn't write (the function body) but I'll dig around to see if we can get the loader to fix this up to the import address. The only complication is that we don't want this to be a pointer to to the entry in the GOT, we want the entry here to be the same as the one in the GOT.

re trap tables, I had imagined import funcs could have trap tables, iff they were imported from a wasm module and we copied their trap information here. But in retrospect that wouldn't be sufficient to be useful for trap lookup (a module's direct dependency's trap info would be present, but not transitive dependencies, if any ever exist), so that really should all be left nil.

What would linkage offer in `function_manifest`? We can learn a function is imported/exported by matching the function index with information in `ModuleData` - are you thinking of a scenario where that wouldn't be workable?

Reply:
I hadn't thought about the GOT issue, hopefully that is resolvable.

Agreed re trap tables.

I only thought the linkage should be available as an alternative to checking `len > 0` when constructing the manifest reloc. It doesn't need to be in the manifest itself.

Reply:
thought about this a bit and realized part of the problem was calling `write_relocated_slice` with constructed trap table symbols that never really existed, expecting `write_relocated_slice` to happen to discard those by relying on implicit knowledge. Rewrote it to make the imported data/local data cases clearer and for callers to not provide junk data.